Issue763 delay blocks

milicag opened the pull request:
This PR introduces two delay blocks required by certain G36 sequences.

- ```OnHold``` - Block that holds a true signal for a defined time period. 
- ```OnOffHold``` - Block that holds an on or off signal constant for at least a defined time period. 

Added to this PR

- ```RealToBoolean``` converter block.

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+    annotation (Placement(transformation(extent={{-100,-40},{-80,-20}})));
+  Timer                                                        timer
+    annotation (Placement(transformation(extent={{20,10},{40,30}})));
+  Modelica.Blocks.Logical.Pre pre
```
Since the `CDL.Logical.Pre` block was added to CDL package, it may be better to change the `Modelica.Blocks.Logical.Pre` to `CDL.Logical.Pre`

milicag replied:
Yes, I rushed to move the ```pre``` block and forgot about that. Will do.

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/Validation/OnHold.mo:
```diff
@@ -0,0 +1,34 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite.Validation;
+model OnHold "Validation model for the OnHold block"
+  import Buildings;
```
delete `import Buildings;`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/Validation/OnOffHold.mo:
```diff
@@ -0,0 +1,35 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite.Validation;
+model OnOffHold "Validation model for the OnOffHold block"
+  import Buildings;
```
delete `import Buildings;`

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Pre.mo:
```diff
@@ -0,0 +1,41 @@
+  "Breaks algebraic loops by an infinitesimal small time delay (y = pre(u): event iteration continues until u = pre(u))"
+
+  parameter Boolean pre_u_start=false "Start value of pre(u) at initial time";
+  extends Modelica.Blocks.Interfaces.partialBooleanSISO;
```
Might it be better to avoid using the `extends` here?

JayHuLBL commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite;
+model OnHold "Block that holds a signal on for a requested time period"
```
Should it be `model` or `block`?

milicag replied:
A block, I'll modify.

milicag commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnOffHold.mo:
```diff
@@ -0,0 +1,141 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite;
```
Fix icon alignment (it's slightly offset from the square grid)

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+
+  LessThreshold                                               les1(threshold=
+        holdOnDuration)
+    annotation (Placement(transformation(extent={{-20,-40},{0,-20}})));
```
Please correct indentation and add comments to each instance.
Also, this assignment should probably be `final` as there is no use to ever overwrite it.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+  Interfaces.BooleanInput u
+    annotation (Placement(transformation(extent={{-140,-20},{-100,20}}),
+        iconTransformation(extent={{-120,-10},{-100,10}})));
+  parameter Real holdOnDuration(unit="s") = 3600 "Time duration of the ON hold.";
```
Use Modelica.SIunits.Time and then the `unit` attribute need not be set.
Also, list first all (public) parameters, then all input signals, then all output signals, and then the instances of variables and blocks.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+    </p>
+    <p>
+    A rising edge of the Boolean input <code>u</code> starts a timer and
+    the Boolean output <code>y</code> stays true until the time
```
Use `<code>true</code>`

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.mo:
```diff
@@ -0,0 +1,102 @@
+
+    <p align=\"center\">
+    <img src=\"modelica://Buildings/Resources/Images/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnHold.PNG\"
+         alt=\"fixme.png\" />
```
Provide an alternate comment, not "fixme.png"

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnOffHold.mo:
```diff
@@ -0,0 +1,141 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite;
+model OnOffHold "The block makes sure that the signal does not change values unless a defined time period has elapsed."
```
Use `block`, not `model`

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Logical/Composite/OnOffHold.mo:
```diff
@@ -0,0 +1,141 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.Logical.Composite;
+model OnOffHold "The block makes sure that the signal does not change values unless a defined time period has elapsed."
+
+  Continuous.Constant Zero(final k=0)
```
Add comments (and correct indentation below)

milicag commented:
@mwetter I've addressed the comments and addressed any additional issues I've noticed during the review. I also added ```RealToBoolean``` converter in this pull request, since it was a single block. My local tests have an issue with the Dymola license, so I'm getting this error:
```
Error: the model is too complex for the current license.
Your license must be upgraded to handle this model.
```
I'll try it again later or wait for Travis to complete the test.

mwetter commented:
This is ready to merge when the CI tests are successful.
